racket/collects/racket/sequence.rkt

```diff
@@ -39,32 +41,56 @@
 (define (sequence->list s)
   (for/list ([v s]) v))
 
-(define (sequence-length s)
-  (unless (sequence? s) (raise-argument-error 'sequence-length "sequence?" s))
-  (cond [(list? s) (length s)]
+(define (sequence-length/o1 s)
```

Comment:
Would it make sense to bail out early when `(struct? s)`. Otherwise, sequence structs (which I'd expect to be a common case) will have to go through all these tests before falling back to the general method. For short sequences, that may matter.